### ISSUE TITLE
Class Cast Exception in aggregation function

### ISSUE THREAD

**Issue by Novemser** [pingcap/tikv-client-java#166]
Known affected branch
```
master
DAGRequestConstruct
``` 

SQL:
```sql
select sum(tp_double),avg(tp_double),min(tp_double),max(tp_double) from full_data_type_table group by tp_double  order by tp_double 
```

TiSpark:
```java
Job aborted due to stage failure: Task 0 in stage 5.0 failed 1 times, most recent failure: Lost task 0.0 in stage 5.0 (TID 601, localhost, executor driver): java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Long
	at scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:105)
	at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow$class.getLong(rows.scala:42)
	at org.apache.spark.sql.catalyst.expressions.GenericInternalRow.getLong(rows.scala:194)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.execution.CoprocessorRDD$$anonfun$doExecute$1$$anonfun$apply$1.apply(CoprocessorRDD.scala:44)
	at org.apache.spark.sql.execution.CoprocessorRDD$$anonfun$doExecute$1$$anonfun$apply$1.apply(CoprocessorRDD.scala:42)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.agg_doAggregateWithKeys$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:377)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:126)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:322)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Spark with JDBC:
```sql
0.00095021725 0.000950217250000 9.5021725E-4 9.5021725E-4 
0.001132071 0.0011320710000 0.001132071 0.001132071 
...dummy out put
```

TiSpark plan:
```bash
== Physical Plan ==
*Project [sum(tp_double)#710, avg(tp_double)#711, min(tp_double)#712, max(tp_double)#713]
+- *Sort [aggOrder#714 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(aggOrder#714 ASC NULLS FIRST, 200)
      +- *HashAggregate(keys=[tp_double#148], functions=[sum(sum(tp_double#148)#715), sum(sum(tp_double#148)#715), sum(count(tp_double#148)#719L), min(min(tp_double#148)#716), max(max(tp_double#148)#717)])
         +- Exchange hashpartitioning(tp_double#148, 200)
            +- *HashAggregate(keys=[tp_double#148], functions=[partial_sum(sum(tp_double#148)#715), partial_sum(sum(tp_double#148)#715), partial_sum(count(tp_double#148)#719L), partial_min(min(tp_double#148)#716), partial_max(max(tp_double#148)#717)])
               +- TiDB CoprocessorRDD{[table: full_data_type_table] , Ranges: Start:[-9223372036854775808], End: [9223372036854775807], Columns: [tp_double], Aggregates: sum([tp_double]), sum([tp_double]), count([tp_double]), min([tp_double]), max([tp_double]), Group By: [[tp_double] ASC]}
```

Spark plan:
```bash
== Physical Plan ==
*Project [sum(tp_double)#324, avg(tp_double)#325, min(tp_double)#326, max(tp_double)#327]
+- *Sort [aggOrder#328 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(aggOrder#328 ASC NULLS FIRST, 200)
      +- *HashAggregate(keys=[tp_double#10], functions=[sum(tp_double#10), avg(tp_double#10), min(tp_double#10), max(tp_double#10)])
         +- Exchange hashpartitioning(tp_double#10, 200)
            +- *HashAggregate(keys=[tp_double#10], functions=[partial_sum(tp_double#10), partial_avg(tp_double#10), partial_min(tp_double#10), partial_max(tp_double#10)])
               +- *Scan JDBCRelation(tispark_test.full_data_type_table) [numPartitions=1] [tp_double#10] ReadSchema: struct<tp_double:double>
```

**Comment by Novemser** [comment]
Looks `HashAggregate` may has some issue.

**Comment by ilovesoup** [comment]
This affects both master and DAG right?
Please take a look to confirm if it indeed caused by duplicate expression elimination.
@Novemser

**Comment by Novemser** [comment]
Branch `master` is also affected.
Caused by duplicate expression elimination.
Should be fixed by pr [#tispark#97](https://github.com/pingcap/tispark/pull/97), which explained why we encountered this issue.
@ilovesoup

**Comment by Novemser** [comment]
Seems not a issue caused by kv client lib but tispark. Should we close this issue and move this to tispark? @ilovesoup
